### DISCUSSION TITLE
Curl handle pool

### DISCUSSION THREAD

**Discussion by hiic** [curl/curl#13773] [General]
Hello, 

In our library we need to implement handle pool. For now using easy interface. To reuse handle/connection, each incoming thread asks for a handle from a pool (under the lock). If a handle is available, a thread takes it away. If a handle is not available, a new one is created. Each use of a handle differs in: 
1. Headers
2. Post message (obviously)
3. Possibly secondary settings, e.g. timeout

Few questions:
1. So once a pool returns a handle, a thread needs to reset headers and reset post message. I want to make sure I do it right (no memory leaks, etc.). I understand simply setting options (with curl_easy_setopt) is the right way to do it? Or do I reset all options first? Or should I reset to nullptr/0 after thread returns handle back to the pool 
2. What the best most efficient way to create a new handle if another one already exists (e.g. curl_easy_duphandle or some other copy). This may happen during the existing  handle is in use by a diff thread. Perhaps we can create a master /template handle that is only used to make working handles (though this would result in having extra connection established)?

Thank you.

**Comment by icing** [comment]
You say you "need" a handle pool, but it is not apparent what you try to achieve with this. Perhaps you can describe what your goals are before we discuss solutions?

As a note: multi-threaded used of handles is not supported. You may run into spurious errors and memory failures when doing so. The general model is that each thread uses its own handles and multihandle. There are possibilities for "sharing" some things between them. What would apply best depends on what you need to do.

**Reply by hiic** [reply]
We have a multiple threads or a single thread runs (depending on the config). Usually multiple threads. Each thread may need to transfer data from remote server. Currently every request always gets a brand new handle (calls curl_easy_init). This is not recommended and inefficient. And yes, we can’t reuse same handle at the same time in two threads. This is exactly we considering a pool. First thread comes in, new handle is created. Second thread comes in, it will get existing handle if T1 is done using it, otherwise new one is created. When a tread is done using a handle, it is retuned back to the pool. So the number of total handles in the pool depends on how threads are asking for it. It is still possible that  all threads ask for them at about same time and each gets a new one. But at least if later a thread needs to transfer again, it will be using existing handle. Of cause “take from pool” and “return to pool” is done under the lock.

**Comment by icing** [comment]
Are you using `curl_easy_perform()` or `curl_multi_perform()` in your application? This makes a huge difference in sharing of initialised data and existing connections. With the former, it is tied to each easy handle and destroy a handle and creating a new one is indeed costly. With the latter, where everything is kept at the multi handle, the creation of an easy handle is relatively cheap.

To figure out the best way forward for your app design, one would need to know more details. If this is a commercial application, I recommend buying support via https://curl.se/support.html to get the most out of your curl use and avoid painful design reversals later on.

**Reply by hiic** [reply]
We use curl_easy_perform. Will most likely use multi but not right now. I work for the bank. Not possible for me to get the full support from them.

**Reply by icing** [reply]
Well, I cannot tell you what makes the most sense for your or your employer. 

Without knowing more, I suspect that the "pool" optimization you try is not worth the effort and that you'll get more out of using the multi handle. That one, done right, can handle many requests in a single thread with gigabytes of throughput.

**Reply by hiic** [reply]
Ok, that makes sense. Will definitely research multi interface as the next step. For now it will be pool of curl easy. It is very basic. Once the handle is created it is placed in the pool. That’s all it is. It is still better than keep creating new ones. The reason I do not want to use multi at this point is because I do not know much about it. I did read docs but not everything is clear, lots of questions about details and meaning of some explanation. Can you recommend a good read about multi?

**Comment by dfandrich** [comment]
There's an overview at
https://curl.se/libcurl/c/libcurl-multi.html
There's also a section on the multi interface in the Mastering libcurl video
series: https://www.youtube.com/watch?v=ZQXv5v9xocU&t=4247s
